### DISCUSSION TITLE
import zustand in HTML file

### DISCUSSION THREAD

**Discussion by thipages** [pmndrs/zustand#1952] [General] [answered]
Hi,

How can I import zustand in an html file like

```javascript
    <script type="module" src=" .... "></script>
```
or 

```javascript
    <script type="importmap">
        {
          "imports": {
            "zustand":   " .... "
          }
        }
    </script>
```

Thank you

**Comment by dai-shi** [comment]
Use https://esm.sh

**Reply by thipages** [reply]
Hi,
thanks for your quick answer
I just tried it in an esm module with a miminal example and it throws an error
any idea? 
Thank you

```javascript
import {create} from "https://esm.sh/zustand/"
console.log(create) // prints e=>p(e)
const useBearStore = create((set) => ({
    bears: 0
  }))
// throws Uncaught TypeError: Cannot read properties of null (reading 'useRef') in react.production.min.js:25
const bears = useBearStore((state) => state.bears)
```

**Reply by dbritto-dev** [reply]
> Hi,
> thanks for your quick answer
> I just tried it in an esm module with a miminal example and it throws an error
> any idea? 
> Thank you
> 
> ```javascript
> import {create} from "https://esm.sh/zustand/"
> console.log(create) // prints e=>p(e)
> const useBearStore = create((set) => ({
>     bears: 0
>   }))
> // throws Uncaught TypeError: Cannot read properties of null (reading 'useRef') in react.production.min.js:25
> const bears = useBearStore((state) => state.bears)
> ```

Yeah, that's correct since you need `react` if you are not using `zustand/vanilla`

**Reply by thipages** [reply]
ok my bad, I am not using react but "vanilla js"
still got an error :-(

```javascript
import create from "https://esm.sh/zustand/vanilla" // note : I took the default export that time
console.log(create) // prints t=>b(t)
const useBearStore = create((set) => ({
    bears: 0
}))
// throws Uncaught TypeError: useBearStore is not a function in index.js:9
const bears = useBearStore((state) => state.bears)
```

**Reply by dbritto-dev** [reply]
> ok my bad, I am not using react but "vanilla js"
> still got an error :-(
> 
> ```javascript
> import create from "https://esm.sh/zustand/vanilla" // note : I took the default export that time
> console.log(create) // prints t=>b(t)
> const useBearStore = create((set) => ({
>     bears: 0
> }))
> // throws Uncaught TypeError: useBearStore is not a function in index.js:9
> const bears = useBearStore((state) => state.bears)
> ```
> 

@thipages here you go https://docs.pmnd.rs/zustand/recipes/recipes#using-zustand-without-react

**Reply by thipages** [reply, marked as answer]
Thank you, this is working fine now. Code example:

```javascript
import { createStore } from 'https://esm.sh/zustand/vanilla'

const store = createStore((set) => ({
    bears: 0
}))
const { getState, setState, subscribe, destroy } = store
setState((state) => ( {bears : state.bears +1} ) )
console.log(getState((state) => state.bears)) // prints {bears: 1}
```
from my original question :
- 🙅  adding a reference  in an html file (in a `script` tag either via `module` or `importmap` type)  is not relevant cause of the dynamic build /external reference of esm.sh.
- 🆗 import can be done directly in an esm module through `import { createStore } from 'https://esm.sh/zustand/vanilla'`

**Comment by thipages** [comment]
solved. Thank you
